### PR TITLE
Bug 1093828 - Add a site-wide banner to warn users that the database is read-only

### PR DESCRIPTION
**Pull request by bobsilverberg**
Update the Javascript so that the banner will appear on every page, even if dismissed.

As discussed in the bug, making the banner disappear permanently can be confusing to users. I updated this so that the banner can be dismissed on the page, but will show up again on the next page. This is how the SUMO staging banner works.

@peterbe r?

### CODE REVIEW

**Comment by peterbe** [comment]
Hmm... I think we want to change the design altogether. Something like this:
screenshot 2015-01-15 13 06 25
I.e. a banner that doesn't hover over the content but that stays there at the top.

File: sass/settings/_sizes.sass [hunk 1]
```diff
@@ -4,40 +4,40 @@ $ratio: 1.5
...
-$xsmall       : ms(-2,$norm)
-$small        : ms(-1,$norm)
-$small-norm   : (ms(-1,$norm) + $norm)/2
-$large-norm   : (ms(1,$norm) + $norm)/2
-$med          : ms(1,$norm)
-$large        : ms(2,$norm)
-$xlarge       : ms(3,$norm)
-$xxlarge      : ms(5,$norm)
-$xxxlarge     : ms(8,$norm)
+$xxxsmall     : ms(-4,$norm,$ratio)
```
Comment (bobsilverberg) [inline sass/settings/_sizes.sass]: I had to make these changes because compass was unable to compile the file otherwise. I was getting the error:

`error sass/screen.sass (Line 7 of sass/settings/_sizes.sass: Function ms is missing argument $ratio.)`

I am not sure whether the value I supplied is correct or not though. Using `$ratio` as set above was just a guess, but the site seems to look the same.

**Comment by bobsilverberg** [comment]
@peterbe I updated the PR to use a permanent banner at the top of the screen, as per your suggestion.

r?

**Comment by peterbe** [comment]
I'm reopening to make the tests start over. For some reason it passed in python 2.6 but not in 2.7!?!?

File: sass/layout/_site.sass [hunk 2]
```diff
@@ -36,47 +36,11 @@ body > div
...
-    cursor: pointer
-    height: 24px
-    left: -8px
-    position: absolute
-    text-indent: -9999px
-    top: -8px
-    width: 24px
+#read-only-message
+  border: 2px solid red
+  color: red
```
Comment (peterbe) [inline sass/layout/_site.sass]: My eyes! they're bleeding! :)

That red is too stark. Can you change both to `darkred`. If you do, it'll look like this:
screenshot 2015-01-16 08 47 22

File: templates/base_site.html [hunk 3]
```diff
@@ -7,19 +7,15 @@
...
-    <p>There is a temporary problem and Moztrap has been placed into a <b>read-only state</b>.<br>
-    You may browse information on the site but changes will not be saved.<br>
-    Please contact us in #moztrap for more information.</p>
-  </div>
-{% endif %}
 
 <header role="banner">
+  {% if DB_READ_ONLY %}
+    <div id="read-only-message">
+      There is a temporary problem and Moztrap has been placed into a <b>read-only state</b>.
```
Comment (peterbe) [inline templates/base_site.html]: I believe it's `MozTrap` not `Moztrap`. Like `GitHub`.

**Comment by peterbe** [comment]
r+ with two nits.

**Comment by bobsilverberg** [comment]
Thanks @peterbe. Sorry about your eyes.

@camd Final r? please and thank you

**Comment by camd** [comment]
I was just running this locally, and I think we may have some layout issues:
screenshot 2015-01-16 12 38 43

screenshot 2015-01-16 12 38 50

I suspect this is due to the ``sass`` issues you were hitting.  I'll try to experiment with this today and see what I can come up with.  SASS is nifty, but sometimes also frustrating...  :-/

**Comment by bobsilverberg** [comment]
@camd It might be worth just setting the ratio to 1, but that's just a shot in the dark. Thanks for looking into it.

**Comment by camd** [comment]
I'm hitting errors getting the compass stuff to compile at all.  I'll try again on Tuesday when we're back.  @bobsilverberg if you're online on Tuesday, perhaps we could try to work through things together.  I wanted to check my package versions against yours.  

I keep getting an error:

    error sass/screen.sass (Line 61 of sass/defaults/_forms.sass: Undefined variable: "$base-rhythm-unit".)

I tried defining that, but then got:

    error sass/screen.sass (Line 80 of sass/defaults/_tags.sass: "pre" failed to @extend ".mono".
    The selector ".mono" was not found.
    Use "@extend .mono !optional" if the extend should be able to fail.
    )

so I think I have some wrong/incomplete versions of files or something...

**Comment by bobsilverberg** [comment]
@camd How are things going with this? I guess the changes I was forced to make to the .sass file are what resulted in this issue. It would have been nice to not have to make those changes, but the css file wouldn't compile without them. :(

I guess ideally we'd find a set of versions that will work with the current sass files in master, but will also still run locally. It's too bad we are dependent on a utility running locally in order to make simple changes to a style, but I guess we're stuck with that.

**Comment by camd** [comment]
@bobsilverberg - Hey Bob.  Unfortunately, I've been slammed last week and this week and just haven't had time to work on this, or anything MozTrap related.  I hope to have some time on Monday, but tomorrow is probably still a wash as well.  

Yeah, I think at this point we're stuck with the sass stuff without a large overhaul.  I found it confusing to work with, though some people just love it.  
Again, sorry for the delays...

**Comment by camd** [comment]
@bobsilverberg - So I finally got around to looking at this.  I had forgotten that, to install the ruby gems, you need to run the script in ``bin/install-gems``  I had to use ``sudo`` to install them.  

However, I got into some crazy dependencies that were really throwing me.  But I finally got the right combo!

So, I first did: 

    sudo gem update --system

This has much nicer ``uninstall`` which lets you target specific versions.  My problem was I had later version of ``chunky_png==1.2.1`` and ``modular-scale==0.0.4`` which kept installing ``sass`` 3.4.11.

To remedy this, first change ``requirements/gems.txt`` to have ``sass==3.2.1``,  Then do the following for all the packages in that file:

    sudo gem uninstall <package>

And delete the versions that are not what ``gems.txt`` is set to.  After that, you can revert your ``$ration`` change and run ``compass watch --poll`` so that any ``.sass`` file updates are reflected in ``screen.css``.

I hope this helps.  And I am infinitely sorry for taking so long to get to this.  Ruby and gems aren't really my thing, so this was a learning experience....

**Comment by camd** [comment]
As a follow up, when uninstalling newer versions, an example would be:

    sudo gem uninstall chunky_png

and then rubygems will let you pick which version to remove.

**Comment by bobsilverberg** [comment]
Hey @camd. Thanks for the above, but I hate to say it didn't work for me. I did what you suggested, but when I try to compile the sass files I am still getting the error:

`error sass/screen.sass (Line 7 of sass/defaults/../settings/_sizes.sass: Function ms is missing argument $ratio.)`

Here is what `gem list` shows me:

```
*** LOCAL GEMS ***

animation (0.1.alpha.2)
bigdecimal (1.2.0)
CFPropertyList (2.2.8)
chunky_png (1.3.3)
compass (0.12.0)
compass-core (1.0.1)
compass-import-once (1.0.5)
compass-susy-plugin (0.9)
ffi (1.9.6)
fssm (0.2.10)
io-console (0.4.2)
json (1.7.7)
libxml-ruby (2.6.0)
minitest (4.3.2)
modular-scale (0.0.4)
multi_json (1.10.1)
nokogiri (1.5.6)
psych (2.0.0)
rake (0.9.6)
rb-fsevent (0.9.4)
rb-inotify (0.9.5)
rdoc (4.0.0)
rubygems-update (2.4.5)
sass (3.2.1)
sqlite3 (1.3.7)
test-unit (2.0.0.0)
```

**Comment by bobsilverberg** [comment]
@camd I noticed some of those versions were off, so I removed all gems and added them individually and then removed any incorrect versions. My `gem list` is now:

```
animation (0.1.alpha.2)
bigdecimal (1.2.0)
CFPropertyList (2.2.8)
chunky_png (1.2.1)
compass (0.12.0)
compass-susy-plugin (0.9)
fssm (0.2.10)
io-console (0.4.2)
json (1.7.7)
libxml-ruby (2.6.0)
minitest (4.3.2)
modular-scale (0.0.4)
nokogiri (1.5.6)
psych (2.0.0)
rake (0.9.6)
rdoc (4.0.0)
sass (3.2.1)
sqlite3 (1.3.7)
test-unit (2.0.0.0)
```

but I am still seeing that same error when trying to compile the sass files.

**Comment by stephendonner** [comment]
Sorry to bug you two, but @peterbe / @camd can you help @bobsilverberg so we can fix, close this out, and resume test-automation runs[1] on moztrap-dev?  Thanks!

[1] https://webqa-ci.mozilla.com/view/MozTrap/

**Comment by camd** [comment]
Dang, sorry @bobsilverberg...   I've been slacking on this again.  And I just won't be able to get to it before I go on PTO tomorrow.  I apologize.  Peterbe, if you have cycles, that'd be awesome.  If not, I'll have to take a look when I get back.

**Comment by peterbe** [comment]
What's the state of things? How did you manage to make the first version of the read-only state thing? Why can't you compile the CSS now that you've changed the design?

What would it take to radically re-do our SCSS dependencies so that things work with modern versions? I think we should take our gloves off and move faster.

**Comment by bobsilverberg** [comment]
@peterbe I believe the first version of the read-only state thing didn't involve any changes to the css, which is why I didn't have to recompile the sass. Once I changed the sass files I needed to recompile and started running into problems. First I was unable to recompile at all because of versions of the ruby gems. After resolving that, I was unable to recompile the current sass files successfully as there seemed to be incompatibilities between the files and my new versions of the gems. I was able to change the sass files in order to get them to recompile, but the resulting css was no longer okay.

The bottom line is that I am unable to make any changes to the css without running into these issues, and I believe that @camd tried as well and had similar issues. This seems like a pretty dangerous position to be in, so I agree that we should try to resolve the issue by coming up with a set of modern versions that work.

**Comment by peterbe** [comment]
So your first version didn't edit any of the existing scss files? Could you not to the same with this second attempt where it's a top-banner (instead of a floating overlay)? Just to get the heck out of this funk :)

> so I agree that we should try to resolve the issue by coming up with a set of modern versions that work.

Ok. So you want to make the scss compile properly with modern (and available) gems? Cool.

**Comment by bobsilverberg** [comment]
@peterbe I was incorrect when I said that the original PR that I submitted didn't change the css. It did, as can be seen from this PR which removes some of those changes. Why I was able to compile those, but not the new ones I cannot say for sure, but my OS was upgraded, which is likely the problem.

I have created a new branch which attempts to make the same changes as this one, but without any css changes. Now I need to try to find some existing css that I can use to format the message the way we want (which may or may not be possible). As you can imagine, this whole thing is a bit of a nightmare. It definitely would be much better if we could find a set of versions that allow us to compile the sass files as they currently stand, but I'm not sure when I can find the time it would take to do that. I imagine that might take a long time with experimenting with different versions.

I am sorry this is taking so long to address. Unfortunately I have many other priorities on my plate and can only find time to work on Moztrap from time to time.

I wonder if we should perhaps abandon this effort for now, and go with the message as was implemented originally, but simply not allow the box to be permanently dismissed? I think that would address all of the current issues, but would be somewhat annoying for users. Having said that, how often do we expect the DB to end up in a read-only state anyway?

I am not suggesting abandoning this forever, but rather getting things into an acceptable stage for now. What do you think?

**Comment by peterbe** [comment]
Cool. 

* We've made it so that you can close the message but not have that stored in localStorage. That was [#91](https://github.com/mozilla/moztrap/pull/91)
* Using modern gems it's currently impossible to build working CSS. :( but it'll have to wait. 
* I think we should stick with the pop-up you made and just move on. Being in read-only mode isn't super common anyway.
